### PR TITLE
Annotation based encryption

### PR DESCRIPTION
**Pull request by nidhi-nair**
Introduces `@Encrypted` as an annotation that works on Mongo listeners. Does not use this annotation in any of the domains yet. Will require all candidates to implement the new `AppsmithDomain` that has been introduced to lower the number of objects analyzed for this annotation.

### CODE REVIEW

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        // Don't bother with primitives
+        if (ClassUtils.isPrimitiveOrWrapper(sourceClass)) return Collections.emptyList();
+
+        // If it is not known, scan each field for annotation or Appsmith type
+        List<CandidateField> finalCandidateFields = new ArrayList<>();
+        ReflectionUtils.doWithFields(sourceClass, field -> {
+            if (field.getAnnotation(Encrypted.class) != null) {
+                CandidateField candidateField = new CandidateField(field, CandidateField.Type.ANNOTATED_FIELD);
+                finalCandidateFields.add(candidateField);
+            } else if (AppsmithDomain.class.isAssignableFrom(field.getType())) {
```
Comment (sharat87) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:51]: Isn't the predicate function on line 96 already doing these checks? Will we still see fields that don't have this annotation here?

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/Encrypted.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * We expect to use this annotation only at a field level.
+ * Since we don't use DBRefs, all fields inside document objects are candidates for encryption
+ * Our implementation of encryption means that ant field marked by this annotation will be encrypted in its entirety
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.FIELD})
+public @interface Encrypted {
+}
```
Comment (sharat87) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/Encrypted.java:16]: Asking because I'm unfamiliar with this API. Is there a way to specify that this annotation can only be used on `String` fields?

Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/Encrypted.java:16]: Nevermind. Looks like you are handling encryption of nested String fields on non-`String` fields. 👍

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/documenttype/DocumentType.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package com.appsmith.external.annotations;
+package com.appsmith.external.annotations.documenttype;
```
Comment (mohanarpit) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/documenttype/DocumentType.java:1]: Super minor: Suggestion to use Camel case. Keeps it consistent.
```suggestion
package com.appsmith.external.annotations.documentType;
```

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionMongoEventListener.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    // This lifecycle event is before we save a document into the DB,
+    // and even before the mapper has converted the object into a document type
+    @Override
+    public void onBeforeConvert(BeforeConvertEvent<E> event) {
+        E source = event.getSource();
+
+        encryptionHandler.convertEncryption(source, encryptionService::encryptString);
+
+//        log.debug("onBeforeConvert for {}", event.getDocument());
```
Comment (mohanarpit) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionMongoEventListener.java:28]: Can we remove these commented debug logs?

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionMongoEventListener.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    public EncryptionMongoEventListener(EncryptionService encryptionService) {
+        encryptionHandler = new EncryptionHandler();
+        this.encryptionService = encryptionService;
+    }
+
+    // This lifecycle event is before we save a document into the DB,
+    // and even before the mapper has converted the object into a document type
+    @Override
+    public void onBeforeConvert(BeforeConvertEvent<E> event) {
```
Comment (mohanarpit) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionMongoEventListener.java:23]: Quick question: I'm wondering if there's going to be an impact on our performance (especially if we retrieve/save larger objects) because of this listener. I'm assuming that we go through the entire object, find the fields that need to be transformed and transform them accordingly right? If there's no real discernible change in performance, it's fine. But I'm curious if you know of the performance impact.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/documenttype/DocumentType.java (hunk 3 above)
Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/documenttype/DocumentType.java:1]: IIRC, the convention is to use lower case for package names always. Since package names are folder names and not all operating systems use case sensitive file names.

Comment (mohanarpit) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/documenttype/DocumentType.java:1]: Aah sorry. My bad. Resolving this comment.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+        // Don't bother with primitives
+        if (ClassUtils.isPrimitiveOrWrapper(sourceClass)) return Collections.emptyList();
+
+        // If it is not known, scan each field for annotation or Appsmith type
+        List<CandidateField> finalCandidateFields = new ArrayList<>();
+        ReflectionUtils.doWithFields(sourceClass, field -> {
+            if (field.getAnnotation(Encrypted.class) != null) {
+                CandidateField candidateField = new CandidateField(field, CandidateField.Type.ANNOTATED_FIELD);
+                finalCandidateFields.add(candidateField);
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:50]: Oh sorry @sharat87, I thought I addressed this in your previous comment. So the two criteria we have is either that the field should be annotated as `Encrypted`, or that the object should be implementing `AppsmithDomain` which will make it a candidate for other encrypted fields inside of it.

Line #95 picks these two types, and here we handle the two cases separately. For `Encrypted`, we simply encrypt/decrypt. For `AppsmithDomain` types, we recurse.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionMongoEventListener.java (hunk 5 above)
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionMongoEventListener.java:23]: There will be a slight difference for sure, but that's only because we will need to traverse all types for candidate fields. We are also using a cache so that identified types can just look up their candidates from the cache.

This means that:
1. After the first few runs of the server the cache will already know about all types.
2. The object size is irrelevant because we are only dealing with types here. The only time we access the value is for encryption/decryption, in which case the overhead is unavoidable.
3. This leaves the reflection overhead, which frankly I haven't measured. I'd tried implementing this for OAuth2 locally, and that didn't impact performance too much on my machine, but I would want to run more rigorous tests once all types have absorbed this mechanism.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:50]: Ah I see, got it. In that case, it looks like the purpose of the `AppsmithDomain` interface is to essentially mark classes as *might-contain-encryptable-fields*. We are using this interface so that we nest inside objects of classes, that we know might contain encrypted fields. Am I getting this right?

If yes, what happens when the field is a composite data type from Java core? Like `List<TypeWithEncryptedFields>`. We don't seem to be traversing through such fields, since they wouldn't pass the predicate on line 95 right? Or are we choosing to solve this problem when we have such fields to deal with (which is okay IMO)?

[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:50]: Please tell me I was addressing the right question and did not misinterpret what you were asking :D

[the remark this answers is not in the document]
Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:50]: You did, actually. My question is more of a follow-up. Likely did a poor job at it 😄. Let me rephrase.

In the predicate on line 95, we are checking if the field's type is assignable from `AppsmithDomain`, and if it is, we are nesting inside that field to find any encrypted fields. (I'm hoping I got this right. If it isn't, disregard the rest of this comment 🤫).

Now, my question is what happens when we encounter a field of type `List<Connection>`, where the class `Connection` has a field with the `Encrypted` annotation? If the type of this field were just `Connection`, we would've nested deeper inside, since `AppsmithDomain.class.isAssignableFrom(field.getType())` would've been `true`. But when the type is `List<Connection>`, I believe it will be `false` and so we won't nest inside. This means the field values inside the `Connection` instances inside that list won't be encrypted, even if they are annotated, correct?

I likely did a poor job again, so we can get on a call later if you're available. Thanks :)

[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:50]: Oh, yeah that's actually a slip up on my end. That's a really, really good catch. I'll probably need to check for type parameters as well in that case. Coming right up :D

[the remark this answers is not in the document]
Comment (sharat87) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/annotations/encryption/EncryptionHandler.java:50]: No problem. I’m guessing this problem doesn’t occur today since we don’t have any such fields. So, I’m okay if we want to solve this another day, when we face it. I just wanted us to be aware.

Let me know if you want to go that way and I’ll approve. Rest of this PR looks good to go.

**Review by sharat87 — APPROVED** [review]

**Comment by nidhi-nair** [comment]
Merging this in, because this doesn't touch UI, or API interfaces
